File: internal/models/service_monitor.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// ProbeType represents the type of service probe
type ProbeType string

const (
	ProbeTypeHTTP ProbeType = "HTTP"
	ProbeTypeTCP  ProbeType = "TCP"
	ProbeTypeICMP ProbeType = "ICMP"
)

// ProbeStrategy represents how to select probe nodes
type ProbeStrategy string

const (
	ProbeStrategyServer  ProbeStrategy = "server"  // Server-side probing
	ProbeStrategyInclude ProbeStrategy = "include" // Include specific nodes
	ProbeStrategyExclude ProbeStrategy = "exclude" // Exclude specific nodes
	ProbeStrategyGroup   ProbeStrategy = "group"   // Probe from a node group
)

// ServiceMonitor represents a service health check configuration
type ServiceMonitor struct {
	BaseModel

	// Basic information
	Name     string    `gorm:"not null" json:"name"`
	Type     ProbeType `gorm:"not null;index" json:"type"`
	Target   string    `gorm:"not null" json:"target"`   // URL/IP:Port/IP
	Interval int       `gorm:"not null" json:"interval"` // Probe interval in seconds
	Timeout  int       `gorm:"not null" json:"timeout"`  // Timeout in seconds
	Enabled  bool      `gorm:"default:true;index" json:"enabled"`

	// Probe node selection strategy
	ProbeStrategy ProbeStrategy `gorm:"default:'server';index" json:"probe_strategy"`      // server/include/exclude/group
	ProbeNodeIDs  string        `gorm:"type:text" json:"probe_node_ids,omitempty"`         // JSON array of node UUIDs
	ProbeGroupName string       `gorm:"type:varchar(100);index" json:"probe_group_name,omitempty"` // Node group name for group strategy
	HostNodeID    *uuid.UUID    `gorm:"type:char(36);index" json:"host_node_id,omitempty"` // Legacy: single executor host (deprecated, use ProbeStrategy)

	// HTTP-specific configuration
	HTTPMethod   string `json:"http_method,omitempty"`                   // GET/POST
	HTTPHeaders  string `gorm:"type:text" json:"http_headers,omitempty"` // JSON map
	HTTPBody     string `gorm:"type:text" json:"http_body,omitempty"`
	ExpectStatus int    `json:"expect_status,omitempty"` // Expected HTTP status code
	ExpectBody   string `json:"expect_body,omitempty"`   // Expected response body substring

	// TCP-specific configuration
	TCPSend   string `json:"tcp_send,omitempty"`   // Data to send
	TCPExpect string `json:"tcp_expect,omitempty"` // Expected response

	// Alert configuration
	NotifyOnFailure   bool `gorm:"default:true" json:"notify_on_failure"`
	FailureThreshold  int  `gorm:"default:3" json:"failure_threshold"`  // Consecutive failures before alert
	RecoveryThreshold int  `gorm:"default:1" json:"recovery_threshold"` // Consecutive successes before recovery

	// Runtime status (not persisted)
	Status        string     `gorm:"-" json:"status,omitempty"` // online/offline/unknown
	LastCheckTime *time.Time `gorm:"-" json:"last_check_time,omitempty"`
	Uptime24h     float64    `gorm:"-" json:"uptime_24h,omitempty"` // 24h uptime percentage

	// Relationships
	HostNode     *HostNode             `gorm:"foreignKey:HostNodeID" json:"-"`
	ProbeResults []ServiceProbeResult  `gorm:"foreignKey:ServiceMonitorID" json:"-"`
	Availability []ServiceAvailability `gorm:"foreignKey:ServiceMonitorID" json:"-"`
	AlertRules   []MonitorAlertRule    `gorm:"foreignKey:TargetID" json:"-"`
}

// TableName specifies the table name for ServiceMonitor
func (ServiceMonitor) TableName() string {
	return "service_monitors"
}

// ServiceStatus represents the health status of a service
type ServiceStatus string

const (
	ServiceStatusGood            ServiceStatus = "Good"
	ServiceStatusLowAvailability ServiceStatus = "LowAvailability"
	ServiceStatusDown            ServiceStatus = "Down"
	ServiceStatusUnknown         ServiceStatus = "Unknown"
)

// GetStatusCode calculates the service status based on uptime percentage
// Uses the same thresholds as Nezha:
// - >= 95%: Good
// - >= 80%: LowAvailability
// - < 80%: Down
func (s *ServiceMonitor) GetStatusCode(uptimePercent float64) ServiceStatus {
	if uptimePercent >= 95.0 {
		return ServiceStatusGood
	} else if uptimePercent >= 80.0 {
		return ServiceStatusLowAvailability
	}
	return ServiceStatusDown
}

// IsHealthy checks if service is healthy (>= 95% uptime)
func (s *ServiceMonitor) IsHealthy(uptimePercent float64) bool {
	return uptimePercent >= 95.0
}

// BeforeCreate validates the model before creation
func (s *ServiceMonitor) BeforeCreate(tx *gorm.DB) error {
	// Call BaseModel's BeforeCreate to generate UUID
	if err := s.BaseModel.BeforeCreate(tx); err != nil {
		return err
	}

	// Set default thresholds
	if s.FailureThreshold <= 0 {
		s.FailureThreshold = 3
	}
	if s.RecoveryThreshold <= 0 {
		s.RecoveryThreshold = 1
	}
	if s.Interval <= 0 {
		s.Interval = 60
	}
	if s.Timeout <= 0 {
		s.Timeout = 5
	}
	return nil
}

File: ui/src/services/service-monitor.ts
import { apiClient } from '@/lib/api-client';

// Service Monitor API Service

export interface ServiceMonitor {
  id: string;
  name: string;
  type: 'HTTP' | 'TCP' | 'ICMP';
  target: string;
  interval: number;
  timeout: number;
  enabled: boolean;

  // Probe strategy
  probe_strategy?: 'server' | 'include' | 'exclude' | 'group';
  probe_node_ids?: string;  // JSON string array
  probe_group_name?: string; // Node group name for group strategy

  // HTTP-specific
  http_method?: string;
  http_headers?: string;
  http_body?: string;
  expect_status?: number;
  expect_body?: string;

  // TCP-specific
  tcp_send?: string;
  tcp_expect?: string;

  // Alert configuration
  notify_on_failure: boolean;
  failure_threshold: number;
  recovery_threshold: number;

  // Runtime status
  status?: 'up' | 'down' | 'degraded' | 'unknown';
  last_check_time?: string;
  uptime_24h?: number;
  created_at: string;
  updated_at: string;
}

export interface ServiceProbeResult {
  id: string;
  service_monitor_id: string;
  host_node_id?: string;  // Node that executed the probe (null for server-side probes)
  timestamp: string;
  success: boolean;
  latency: number;
  error_message?: string;
  http_status_code?: number;
  http_response_body?: string;
  tcp_response?: string;
}

export interface ServiceMonitorListParams {
  search?: string;
  type?: string;
  status?: string;
  enabled?: boolean;
  page?: number;
  page_size?: number;
}

export interface ServiceMonitorFormData {
  name: string;
  type: 'HTTP' | 'TCP' | 'ICMP';
  target: string;
  interval: number;
  timeout: number;
  enabled: boolean;

  // Probe strategy
  probe_strategy?: 'server' | 'include' | 'exclude' | 'group';
  probe_node_ids?: string;  // JSON string array
  probe_group_name?: string; // Node group name for group strategy

  // HTTP-specific
  http_method?: string;
  http_headers?: string;
  http_body?: string;
  expect_status?: number;
  expect_body?: string;

  // TCP-specific
  tcp_send?: string;
  tcp_expect?: string;

  // Alert configuration
  notify_on_failure: boolean;
  failure_threshold: number;
  recovery_threshold: number;
}

export interface ServiceMonitorListResponse {
  items: ServiceMonitor[];
  total: number;
}

export const ServiceMonitorService = {
  // List service monitors
  async list(params?: ServiceMonitorListParams): Promise<ServiceMonitorListResponse> {
    const queryParams = new URLSearchParams();
    if (params?.search) queryParams.append('search', params.search);
    if (params?.type) queryParams.append('type', params.type);
    if (params?.status) queryParams.append('status', params.status);
    if (params?.enabled !== undefined) queryParams.append('enabled', String(params.enabled));
    if (params?.page) queryParams.append('page', String(params.page));
    if (params?.page_size) queryParams.append('page_size', String(params.page_size));

    const queryString = queryParams.toString();
    const response = await apiClient.get<{
      data?: {
        items?: ServiceMonitor[];
        total?: number;
      };
    }>(
      `/vms/service-monitors${queryString ? `?${queryString}` : ''}`
    );
    return {
      items: response.data?.items || [],
      total: response.data?.total || 0,
    };
  },

  // Get single service monitor
  async get(id: string): Promise<ServiceMonitor> {
    const response = await apiClient.get<{ data: ServiceMonitor }>(`/vms/service-monitors/${id}`);
    return response.data;
  },

  // Create service monitor
  async create(data: ServiceMonitorFormData): Promise<ServiceMonitor> {
    const response = await apiClient.post<{ data: ServiceMonitor }>('/vms/service-monitors', data);
    return response.data;
  },

  // Update service monitor
  async update(id: string, data: Partial<ServiceMonitorFormData>): Promise<ServiceMonitor> {
    const response = await apiClient.put<{ data: ServiceMonitor }>(`/vms/service-monitors/${id}`, data);
    return response.data;
  },

  // Delete service monitor
  async delete(id: string): Promise<void> {
    await apiClient.delete(`/vms/service-monitors/${id}`);
  },

  // Trigger manual probe
  async triggerProbe(id: string): Promise<ServiceProbeResult> {
    const response = await apiClient.post<{ data: ServiceProbeResult }>(
      `/vms/service-monitors/${id}/trigger`
    );
    return response.data;
  },

  // Get probe history
  async getProbeHistory(
    id: string,
    params?: {
      start_time?: string;
      end_time?: string;
      limit?: number
    }
  ): Promise<ServiceProbeResult[]> {
    const queryParams = new URLSearchParams();
    if (params?.start_time) queryParams.append('start_time', params.start_time);
    if (params?.end_time) queryParams.append('end_time', params.end_time);
    if (params?.limit) queryParams.append('limit', String(params.limit));

    const queryString = queryParams.toString();
    const response = await apiClient.get<{ data?: ServiceProbeResult[] }>(
      `/vms/service-monitors/${id}/history${queryString ? `?${queryString}` : ''}`
    );
    return response.data || [];
  },

  // Get availability stats
  async getAvailabilityStats(
    id: string,
    params?: {
      period?: '1h' | '12h' | '24h' | '7d' | '30d';
    }
  ): Promise<{
    uptime_percentage: number;
    total_checks: number;
    successful_checks: number;
    failed_checks: number;
    avg_latency: number;
    min_latency: number;
    max_latency: number;
  }> {
    const queryParams = new URLSearchParams();
    if (params?.period) queryParams.append('period', params.period);

    const queryString = queryParams.toString();
    const response = await apiClient.get<{
      data: {
        uptime_percentage: number;
        total_checks: number;
        successful_checks: number;
        failed_checks: number;
        avg_latency: number;
        min_latency: number;
        max_latency: number;
      };
    }>(`/vms/service-monitors/${id}/availability${queryString ? `?${queryString}` : ''}`);
    return response.data;
  },

  // Batch enable/disable monitors
  async batchUpdate(
    ids: string[],
    action: 'enable' | 'disable'
  ): Promise<void> {
    await apiClient.post('/vms/service-monitors/batch', {
      ids,
      action,
      enabled: action === 'enable',
    });
  },

  // Export monitors configuration
  async export(format: 'json' | 'yaml' = 'json'): Promise<Blob> {
    return apiClient.getBlob('/vms/service-monitors/export', { format });
  },

  // Import monitors configuration
  async import(file: File): Promise<{ imported: number; failed: number }> {
    const formData = new FormData();
    formData.append('file', file);
    const response = await apiClient.postForm<{ data: { imported: number; failed: number } }>(
      '/vms/service-monitors/import',
      formData
    );
    return response.data;
  },

  // Get overview with 30-day statistics for all service monitors
  async getOverview(): Promise<ServiceOverviewResponse> {
    const response = await apiClient.get<{ data: ServiceOverviewResponse }>(
      '/vms/service-monitors/overview'
    );
    return response.data;
  },

  // Get probe history for a specific host (multi-line chart data)
  async getHostProbeHistory(
    hostId: string,
    hours: number = 24
  ): Promise<ServiceHistoryInfo[]> {
    const response = await apiClient.get<{ data?: ServiceHistoryInfo[] }>(
      `/vms/hosts/${hostId}/probe-history`,
      { hours }
    );
    return response.data || [];
  },

  // Get network topology matrix data
  async getNetworkTopology(hours: number = 1): Promise<NetworkTopologyResponse> {
    const response = await apiClient.get<{ data: NetworkTopologyResponse }>(
      '/vms/service-monitors/topology',
      { hours }
    );
    return response.data;
  },
};

// 30-day service statistics response
export interface ServiceResponseItem {
  service_monitor_id: string;
  service_name: string;

  // 30-day arrays (index 0 = today, 29 = 30 days ago)
  delay: number[];       // Average delay per day (ms)
  up: number[];          // Successful checks per day
  down: number[];        // Failed checks per day

  // Aggregated statistics
  total_up: number;
  total_down: number;
  uptime_percentage: number;

  // Current status (from today)
  current_up: number;
  current_down: number;

  // Status code: Good(>95%) / LowAvailability(80-95%) / Down(<80%)
  status_code: 'Good' | 'LowAvailability' | 'Down' | 'Unknown';
}

export interface ServiceOverviewResponse {
  services: Record<string, ServiceResponseItem>; // Key: service_monitor_id
}

// Host probe history for multi-line chart
export interface ServiceHistoryInfo {
  service_monitor_id: string;
  service_monitor_name: string; // Target name
  service_monitor_type: string; // HTTP, TCP, or ICMP
  host_node_id: string;
  host_node_name?: string;      // Executor name
  timestamps: number[];         // Unix timestamps in milliseconds
  avg_delays: number[];         // Average delays in milliseconds
  uptimes: number[];            // Uptime percentages
}

// Network topology types
export interface NetworkTopologyNode {
  id: string;
  name: string;
  type: 'host' | 'service';
  is_online: boolean;
}

export interface NetworkTopologyEdge {
  source_id: string;
  target_id: string;
  avg_latency: number;
  min_latency: number;
  max_latency: number;
  packet_loss: number;
  success_rate: number;
  probe_count: number;
  last_probe_time: string;
}

export interface NetworkTopologyResponse {
  nodes: NetworkTopologyNode[];
  edges: NetworkTopologyEdge[];
  matrix: Record<string, Record<string, NetworkTopologyEdge>>;
}
